Sources/SwiftSyntax/SourceLocation.swift

```diff
@@ -0,0 +1,344 @@
+
+/// Represents a start and end location in a Swift file.
+public struct SourceRange: Codable {
+  /// The beginning location in the source range.
+  public let start: SourceLocation
+
+  /// The beginning location in the source range.
+  public let end: SourceLocation
+
+  public init(start: SourceLocation, end: SourceLocation) {
+    self.start = start
+    self.end = end
+  }
+}
+
+/// Converts `AbsolutePosition`s of syntax nodes to `SourceLocation`s, and
+/// vice-versa. The `AbsolutePosition`s must be originating from nodes that are
+/// part of the same tree that was used to initialize this class.
+public final class SourceLocationConverter {
+  let file: String
+  /// Array of lines and the position at the start of the line.
+  let lines: [AbsolutePosition]
+  /// Position at end of file.
+  let endOfFile: AbsolutePosition
+
+  /// - Parameters:
+  ///   - file: The file path associated with the syntax tree.
+  ///   - tree: The syntax tree to convert positions to line/columns for.
+  public init(file: String, tree: SourceFileSyntax) {
+    self.file = file
+    (self.lines, endOfFile) = computeLines(tree: tree)
+    assert(tree.byteSize == endOfFile.utf8Offset)
+  }
+
+  /// Convert a `AbsolutePosition` to a `SourceLocation`. If the position is
+  /// exceeding the file length then the `SourceLocation` for the end of file
+  /// is returned. If position is negative the location for start of file is
+  /// returned.
+  public func location(for origpos: AbsolutePosition) -> SourceLocation {
+    // Clamp the given position to the end of file if needed.
+    let pos = min(origpos, endOfFile)
+    if pos.utf8Offset < 0 {
+      return SourceLocation(line: 1, column: 1, offset: 0, file: self.file)
+    }
+
+    assert(!lines.isEmpty)
+    var first = lines.startIndex
+    var i: Int
+    var step: Int
+    var count = lines.endIndex - first
+    // Do an upper bound search, first element that is > value. This provides
+    // the line index coming after the one where the position belongs to.
+    while count > 0 {
+      step = count / 2
+      i = first + step
+      if !(pos < lines[i]) {
+        first = i + 1
+        count -= step + 1
+      } else {
+        count = step
+      }
+    }
+
+    assert(first > 0)
+    let lineIdx = first-1
+    let lineStartOffset = lines[lineIdx].utf8Offset
+    let colOffset = pos.utf8Offset - lineStartOffset
+
+    let line = lineIdx+1
+    let column = colOffset+1
+
+    return SourceLocation(line: line, column: column,
+      offset: pos.utf8Offset, file: self.file)
+  }
+
+  /// Convert a line/column to a `SourceLocation`. If the line/column exceeds
+  /// the boundaries of the file or the line, the position returned is one
+  /// adjusted to the closest boundary (beginning/end of file or line).
+  ///
+  /// - Parameters:
+  ///   - line: A 1-based line number.
+  ///   - column: A 1-based, UTF8 offset from the start of line.
+  public func positionFor(line: Int, column: Int) -> AbsolutePosition {
```

Comment:
Nitpick: This might be better as `position(ofLine:column:)`

Reply:
Done.